### ISSUE TITLE
Type stability for GPs

### ISSUE THREAD

**Issue by kaandocal** [JuliaGaussianProcesses/AbstractGPs.jl#155]
It seems like AbstractGPs is not always type stable in the way we'd expect when calling `posterior`:

```julia
using AbstractGPs

gp = GP(Matern32Kernel())

X1 = ColVecs(rand(1, 10))
Y1 = rand(10) 

gp_post = posterior(gp(X1, 0.01), Y1)

X2 = ColVecs(rand(1, 10))
Y2 = rand(10)

typeof(posterior(gp_post(X2, 0.01), Y2)) == typeof(gp_post) # returns false
```

**Comment by devmotion** [comment]
Shouldn't you use
```julia
typeof​(​posterior​(​gp​(X2, ​0.01​), Y2))
```
in your comparison? 

BTW your comparison doesn't guarantee type stability, you would want to check if the type can be inferred correctly. This can be done e.g. with `@code_warntype` and `@inferred`.

**Comment by kaandocal** [comment]
My fault, I forgot the second `typeof` (the result doesn't change however). Just to clarify, by type stability I meant that adding more observations to a GP changes its type in this case!

**Comment by devmotion** [comment]
I can't check it right now but I assume that the problem is a missing `vcat` definition for `ColVecs`, so the type changes when concatenating `X1` and `X2`. Does it work with regular vectors for `X1` and `X2`?

**Comment by kaandocal** [comment]
That is indeed the problem! For 1-dimensional vectors `X1` and `X2` it works. I'll create a PR in KernelFunctions.jl.

**Comment by willtebbutt** [comment]
@kaandocal is this now resolved as #291 is merged?

**Comment by devmotion** [comment]
I'll take that as a yes :slightly_smiling_face:
